drivers/network/netio/netio.c

```diff
@@ -0,0 +1,896 @@
+          _Reserved_ ULONG Flags,
+          _In_opt_ PSOCKADDR RemoteAddress,
+          _In_ ULONG ControlInfoLength,
+          _In_reads_bytes_opt_(ControlInfoLength) PCMSGHDR ControlInfo, _Inout_ PIRP Irp)
+{
+    PIRP tdiIrp = NULL;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    PTDI_CONNECTION_INFORMATION TargetConnectionInfo;
+    NTSTATUS status;
+    void *BufferData;
+    struct NetioContext *nc;
+
+    if (DummyDeviceObject == NULL)
+    {
+        DbgPrint("DummyDeviceObject is NULL, was the DriverEntry funtion called?\n");
+        Irp->IoStatus.Status = STATUS_INVALID_PARAMETER;
+        return STATUS_INVALID_PARAMETER;        /* TODO: something more meaningful */
+    }
+
+    /* Call ourselves. Sets status to pending. The interesting
+     * part happens later.
+     */
+
+    DummyCallDriver(Irp);
+
+    nc = ExAllocatePoolWithTag(NonPagedPool, sizeof(*nc), 'NEIO');
+    if (nc == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    nc->socket = s;
+    nc->UserIrp = Irp;
+
+    TargetConnectionInfo = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (TargetConnectionInfo == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    BufferData = MmGetSystemAddressForMdlSafe(Buffer->Mdl, NormalPagePriority);
+    if (BufferData == NULL)
+    {
+        DbgPrint("Error mapping MDL\n");
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Some drivers (like WinDRBD) do not set this.
+     * But ReactOS seems to require this (else BSOD on completion).
+     */
+
+    if (Irp->Tail.Overlay.Thread == NULL)
+    {
+        Irp->Tail.Overlay.Thread = PsGetCurrentThread();
+    }
+    IoMarkIrpPending(Irp);
+    SocketGet(s);
+    /* TODO: protect by spinlock ... */
+    /* TODO: still needed? */
+    InsertTailList(&s->PendingUserIrps, &Irp->Tail.Overlay.ListEntry);
+
+    /* This will create a tdiIrp: */
+    status =
+        TdiSendDatagram(&tdiIrp, s->LocalAddressFile, ((char *)BufferData) + Buffer->Offset, Buffer->Length,
+                        TargetConnectionInfo, NetioComplete, nc);
+
+    /* TODO: check this again..I would assume if the TDI helper function
+       returns any error, the completion function is not called.
+     */
+    if (!NT_SUCCESS(status))
+    {
+        SocketPut(s);
+        ExFreePoolWithTag(nc, 'NEIO');
+    }
+    return status;
+}
+
+static NTSTATUS WSKAPI
+WskReceiveFrom(_In_ PWSK_SOCKET Socket,
+               _In_ PWSK_BUF Buffer,
+               _Reserved_ ULONG Flags,
+               _Out_opt_ PSOCKADDR RemoteAddress,
+               _Inout_ PULONG ControlLength,
+               _Out_writes_bytes_opt_(*ControlLength) PCMSGHDR ControlInfo,
+               _Out_opt_ PULONG ControlFlags, _Inout_ PIRP Irp)
```

Comment:
```suggestion
WskReceiveFrom(
    _In_ PWSK_SOCKET Socket,
    _In_ PWSK_BUF Buffer,
    _Reserved_ ULONG Flags,
    _Out_opt_ PSOCKADDR RemoteAddress,
    _Inout_ PULONG ControlLength,
    _Out_writes_bytes_opt_(*ControlLength) PCMSGHDR ControlInfo,
    _Out_opt_ PULONG ControlFlags, _Inout_ PIRP Irp)
```
Don't let the indent depend on the function-names length.